In [1]:
# Importing Packages
import numpy as np
from numpy import  nan
import pandas as pd

# Import Dataset
heart_df = pd.read_csv('heart_failure_clinical_records_dataset.csv')

In [2]:
# 1.a
# Print the shape of the dataset
print(heart_df.shape)

(299, 13)


In [3]:
# 1.b
# Print the axes of the dataset
print(heart_df.axes)

[RangeIndex(start=0, stop=299, step=1), Index(['age', 'anaemia', 'creatinine_phosphokinase', 'diabetes',
       'ejection_fraction', 'high_blood_pressure', 'platelets',
       'serum_creatinine', 'serum_sodium', 'sex', 'smoking', 'time',
       'DEATH_EVENT'],
      dtype='object')]


In [4]:
# 1.c
# Print the Dtype of the dataset
print(heart_df.dtypes)

age                         float64
anaemia                       int64
creatinine_phosphokinase      int64
diabetes                      int64
ejection_fraction             int64
high_blood_pressure           int64
platelets                   float64
serum_creatinine            float64
serum_sodium                  int64
sex                           int64
smoking                       int64
time                          int64
DEATH_EVENT                   int64
dtype: object


In [5]:
# 1.e
# Print the first ten rows of the dataset
print(heart_df.head(10))


    age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0  75.0        0                       582         0                 20   
1  55.0        0                      7861         0                 38   
2  65.0        0                       146         0                 20   
3  50.0        1                       111         0                 20   
4  65.0        1                       160         1                 20   
5  90.0        1                        47         0                 40   
6  75.0        1                       246         0                 15   
7  60.0        1                       315         1                 60   
8  65.0        0                       157         0                 65   
9  80.0        1                       123         0                 35   

   high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                    1  265000.00               1.9           130    1   
1                    0  26

** 1.c What is each row of heart_df dataframe termed within the machine learning parlance?*
<br>
A row in for machine learning represents a single observation along with the valus recorded during said obeservation.


** 1.d Based on your knowledge of the dataset and its stated usage, do you think we are dealing with an unsupervised learning problem or a supervised learning problem? Justify your answer*
<br>
We are dealing with a supervised learning problem. The dataset is clearly labeled that is to help categorize the values observed in each entry.

** 1.e How many independent vairables (features,attrivbutes,predictors, etc.) does this dataset have?List down the names of these varaibles*
<br>
* age
* creatinine_phosphokinase
* ejection_fraction
* platelets
* serum_creatine
* serum_sodium
* anaemia
* diabetes
* high_blood_pressure
* sex
* smoking


** 1.f How many dependent variables(if any) does this dataset have? List down the names of these variables.*
* DEATH_EVENT

** 1.g What type of encoding do the categorical variables in the dataset follow*

In [6]:
# 1.h
deceased = heart_df.DEATH_EVENT.value_counts()
print("Samples of Remaining (0) and Deceased (1): ")
print(deceased)

Samples of Remaining (0) and Deceased (1): 
0    203
1     96
Name: DEATH_EVENT, dtype: int64


In [7]:
# 1.i
sex = heart_df.sex.value_counts()
print("Samples of Females (0) and Males (1): ")
print(sex)

Samples of Females (0) and Males (1): 
1    194
0    105
Name: sex, dtype: int64


In [8]:
# 1.j
smoking = heart_df.smoking.value_counts()
print("Samples of Non-Smokers (0) and Smokers (1): ")
print(smoking)

Samples of Non-Smokers (0) and Smokers (1): 
0    203
1     96
Name: smoking, dtype: int64


<b>2. Compute pairwise correlations between variables in the dataset using pandas.DataFrame.corr() function.<b>

In [9]:
heart_df.corr()

age   anaemia  creatinine_phosphokinase  \
age                       1.000000  0.088006                 -0.081584   
anaemia                   0.088006  1.000000                 -0.190741   
creatinine_phosphokinase -0.081584 -0.190741                  1.000000   
diabetes                 -0.101012 -0.012729                 -0.009639   
ejection_fraction         0.060098  0.031557                 -0.044080   
high_blood_pressure       0.093289  0.038182                 -0.070590   
platelets                -0.052354 -0.043786                  0.024463   
serum_creatinine          0.159187  0.052174                 -0.016408   
serum_sodium             -0.045966  0.041882                  0.059550   
sex                       0.065430 -0.094769                  0.079791   
smoking                   0.018668 -0.107290                  0.002421   
time                     -0.224068 -0.141414                 -0.009346   
DEATH_EVENT               0.253729  0.066270                  0.062728   

                          diabetes  ejection_fraction  high_blood_pressure  \
age                      -0.101012           0.060098             0.093289   
anaemia                  -0.012729           0.031557             0.038182   
creatinine_phosphokinase -0.009639          -0.044080            -0.070590   
diabetes                  1.000000          -0.004850            -0.012732   
ejection_fraction        -0.004850           1.000000             0.024445   
high_blood_pressure      -0.012732           0.024445             1.000000   
platelets                 0.092193           0.072177             0.049963   
serum_creatinine         -0.046975          -0.011302            -0.004935   
serum_sodium             -0.089551           0.175902             0.037109   
sex                      -0.157730          -0.148386            -0.104615   
smoking                  -0.147173          -0.067315            -0.055711   
time                      0.033726           0.041729            -0.196439   
DEATH_EVENT              -0.001943          -0.268603             0.079351   

                          platelets  serum_creatinine  serum_sodium       sex  \
age                       -0.052354          0.159187     -0.045966  0.065430   
anaemia                   -0.043786          0.052174      0.041882 -0.094769   
creatinine_phosphokinase   0.024463         -0.016408      0.059550  0.079791   
diabetes                   0.092193         -0.046975     -0.089551 -0.157730   
ejection_fraction          0.072177         -0.011302      0.175902 -0.148386   
high_blood_pressure        0.049963         -0.004935      0.037109 -0.104615   
platelets                  1.000000         -0.041198      0.062125 -0.125120   
serum_creatinine          -0.041198          1.000000     -0.189095  0.006970   
serum_sodium               0.062125         -0.189095      1.000000 -0.027566   
sex                       -0.125120          0.006970     -0.027566  1.000000   
smoking                    0.028234         -0.027414      0.004813  0.445892   
time                       0.010514         -0.149315      0.087640 -0.015608   
DEATH_EVENT               -0.049139          0.294278     -0.195204 -0.004316   

                           smoking      time  DEATH_EVENT  
age                       0.018668 -0.224068     0.253729  
anaemia                  -0.107290 -0.141414     0.066270  
creatinine_phosphokinase  0.002421 -0.009346     0.062728  
diabetes                 -0.147173  0.033726    -0.001943  
ejection_fraction        -0.067315  0.041729    -0.268603  
high_blood_pressure      -0.055711 -0.196439     0.079351  
platelets                 0.028234  0.010514    -0.049139  
serum_creatinine         -0.027414 -0.149315     0.294278  
serum_sodium              0.004813  0.087640    -0.195204  
sex                       0.445892 -0.015608    -0.004316  
smoking                   1.000000 -0.022839    -0.012623  
time                     -0.022839  1.000000  

** 2.a What two variables are the most positively correlated with the DEATH_EVENT variable?*
* Age 0.253729
* Serum_creatinine 0.294278

** 2.b What two variables are the most negatively correlated with the DEATH_EVENT variable?*
* Time -0.526964
* Ejection_fraction -0.268603

** 2.c Based on your knowledge of the dataset, why do you think it makes sense that the second-most positively correlated variables with the DEATH_EVENT variable should have been positively correlated?*

The second-highest correlation in the dataset with the DEATH_EVENT is serum_creatinine. Serum_creatinine is an indicator of kidney failure. Kidney failure is a major precursor to heart failure.

** 2.d Based on your knowledge of the dataset, why do you think it makes sense that the two most negatively correlated variables with the DEATH_EVENT variable should have been negatively correlated?*

The Time correlation is unclear how it affects the outcome. For the other ejection_fraction, it can be low but not necessarily mean that there is an issue with heart failure. Which in turn does not mean there is heart failure present.

In [10]:
# 3

# Validate data by type and corresponding values
# If there is an error with the value convert said value to NaN

# Two types of values
# 1 Binary (1 yes 0 no)
#  - anaemia
#  - diabetes
#  - high blood pressure
#  - sex (1 male 0 female)
#  - smoking
#  - Death Event

# 2 Ints (0 -> inf)
#  - ejection fraction
#  - platelets
#  - serum creatinine
#  - serum sodium
#  - time

# Validate binary values ( values has to be either 0 or 1 if not convert to NaN)
# for i in range(len(heart_df)):
#     # Check anaemia
#     if heart_df.anaemia[i] < 0 or heart_df.anaemia[i] > 1:
#             heart_df.anaemia[i] = float=("NaN")
#     # Check diabetes
#     if heart_df.diabetes[i] < 0 or heart_df.diabetes[i] > 1:
#         heart_df.diabetes[i] = float=("NaN")
#     # Check high blood pressure
#     if heart_df.high_blood_pressure[i] < 0 or heart_df.high_blood_pressure[i] > 1:
#         heart_df.high_blood_pressure[i] = float=("NaN")
#     # sex
#     if heart_df.sex[i] < 0 or heart_df.sex[i] > 1:
#         heart_df.sex[i] = float=("NaN")
#     # Check smoking
#     if heart_df.smoking[i] < 0 or heart_df.smoking[i] > 1:
#         heart_df.smoking[i] = float=("NaN")
#     # Check Death Event
#     if heart_df.DEATH_EVENT[i] < 0 or heart_df.DEATH_EVENT[i] > 1:
#         heart_df.DEATH_EVENT[i] = float=("NaN")
#
# # Any values in the dataset that are < 0 replace with NaN
# for i in range(len(heart_df)):
#     # Check age
#     if heart_df.age[i] <= 0:
#         heart_df.age[i] = float("NaN")
#     # Check creatinine phosphokinase
#     if heart_df.creatinine_phosphokinase[i] <= 0:
#         heart_df.creatinine_phosphokinase[i] = float("NaN")
#     # Check ejection fraction
#     if heart_df.ejection_fraction[i] < 0 or heart_df.ejection_fraction[i] > 100:
#         heart_df.ejection_fraction[i] = float("NaN")
#     # Check platelets
#     if heart_df.platelets[i] <= 0:
#         heart_df.platelets[i] = float("NaN")
#     # Check serum creatinine
#     if heart_df.serum_creatinine[i] <= 0:
#         heart_df.serum_creatinine[i] = float("NaN")
#     # Check serum sodium
#     if heart_df.serum_sodium[i] <= 0:
#         heart_df.serum_sodium[i] = float("NaN")
#     # Check time
#     if heart_df.time[i] <= 0:
#         heart_df.time[i] = float("NaN")
#
#
# # Write the dataframe with NaN into a new file
# heart_df.to_csv('heart_failure_dataset_NaNs.csv', index=False)

My thinking for the dataset validation is pretty simple. Using the website I used to download the along with the dtypes
I was able to produce for the dataset, I have broken down which dtypes are supposed to be binary values or which ones
are supposed to be natural numbers. For the binary if the value does not correspond to either a 0 or a 1 that value is
not valid and can be replaced with NaN. For the other set, natural numbers are numbers that are positive from 0 to
infinity. So the program checks to make sure there are no negative numbers. If the program finds a negative number,
the value is replaced with a NaN

In [11]:
# 4
# heart_df = heart_df/heart_df.loc[heart_df.abs().idxmax()].astype(np.float64)
# print(heart_df.head(20))

    age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0   1.0      NaN                       1.0       NaN                1.0   
0   1.0      NaN                       1.0       NaN                1.0   
0   1.0      NaN                       1.0       NaN                1.0   
1   1.0      NaN                       1.0       NaN                1.0   
2   1.0      NaN                       1.0       NaN                1.0   
3   1.0      1.0                       1.0       NaN                1.0   
4   1.0      1.0                       1.0       1.0                1.0   
5   NaN      NaN                       NaN       NaN                NaN   
6   NaN      NaN                       NaN       NaN                NaN   
7   NaN      NaN                       NaN       NaN                NaN   
8   NaN      NaN                       NaN       NaN                NaN   
9   1.0      1.0                       1.0       NaN                1.0   
10  NaN      NaN         

In [12]:
# 5
# One-Hot encoded values of the DEATH_EVENT
# x = pd.get_dummies(heart_df.DEATH_EVENT, prefix='DEATH_EVENT')
# # Save to csv file
# x.to_csv('heart_failure_dataset_processed.csv', index=False)
